### PR TITLE
Documentation improvements

### PR DESCRIPTION
**Pull request by devos50**
In this PR, I made various improvements to the (outdated) documentation:
- Updated REST API docs, removed non-existing endpoints and included new endpoints. For our convenience, we should have a check that automatically includes these endpoints.
- Fixed all warnings when building the documentation.
- Added doc build status badge to README.md.
- Added [a PR job](https://jenkins-ci.tribler.org/job/GH_Tribler_PR_Tests/job/PR_build_docs/) that builds the documentation and fails when the documentation cannot be built successfully.

Note that I consider the extensive mocking in `conf.py` not pretty, but necessary since we cannot build dependencies like `libtorrent` in the `virtualenv` environment of readthedocs.

### CODE REVIEW

**Comment by devos50** [comment]
retest this please

**Comment by heldersepu** [comment]
Would be nice if the REST API docs was using swagger/OAS

**Comment by heldersepu** [comment]
I could not find this section:
https://github.com/Tribler/tribler/blob/devel/Tribler/Core/Modules/restapi/downloads_endpoint.py#L444
anywhere in the documentation

**Comment by qstokkink** [comment]
> 
> 
> Would be nice if the REST API docs was using swagger/OAS

We are currently creating a proof of concept using Swagger. It will take a little while for this to make it into Tribler though.

**Comment by Dmole** [comment]
How about "Self-documenting code"?

File: README.rst [hunk 1]
```diff
@@ -2,11 +2,11 @@
 Tribler
 *******
 
-|jenkins_build| |contributors| |pr_closed| |issues_closed| |downloads_7_0| |downloads_7_1| |downloads_7_2| |doi| |openhub|
+|jenkins_build| |docs| |contributors| |pr_closed| |issues_closed| |downloads_7_0| |downloads_7_1| |downloads_7_2| |doi| |openhub|
 
 *Towards making Bittorrent anonymous and impossible to shut down.*
 
-Developers usually hang out in the official IRC channel #tribler @ FreeNode (click `here <http://webchat.freenode.net/?channels=tribler>`_ for direct a webchat window)
+Developers usually hang out in the official IRC channel #tribler @ FreeNode (click `here <http://webchat.freenode.net/?channels=tribler>`__ for direct a webchat window)
```
Comment (ichorid) [inline README.rst]: Except they really don't... :wink:

File: Tribler/Core/Modules/restapi/search_endpoint.py [hunk 2]
```diff
@@ -16,8 +16,8 @@
 
 class SearchEndpoint(BaseMetadataEndpoint):
     """
-    This endpoint is responsible for searching in channels and torrents present in the local Tribler database. It also
-    fires a remote search in the Dispersy communities.
+    This endpoint is responsible for searching in channels and torrents present in the local Tribler database.
+    It also fires a remote search in the IPv8 channel community.
```
Comment (ichorid) [inline Tribler/Core/Modules/restapi/search_endpoint.py]: Shouldn't this be "GigaChannel" community instead?

**Comment by Dmole** [comment]
I think the target for API documentation would be a power user trying to implement some automation without having to waste time reading 3994+ lines of code, when it could be reduced to ~81 lines with some trivial 4 line script;
```
while read -r F ; do
	echo "$F" | perl -pe 's/.*\///g;s/_.*//g'
	grep  -P "in (request.args|parameters)" "$F" | perl -pe 's/if //g;s/^[ \t]*/\t\t/g;s/ in .*//g'
done < <(find Tribler/Core/Modules/restapi/ -name '*_endpoint.py')
```

**Comment by qstokkink** [comment]
> 
> 
> How about "Self-documenting code"?
> (Just have a script build the documentation from the API code)

This is exactly why we need something like Swagger.

Currently all our REST endpoints are custom magic, which cannot be automatically read ([I did, at one point, document the REST API in our networking library using a custom Swagger-like specification to deal with our own magic code](https://github.com/Tribler/py-ipv8/files/2708742/output.pdf)).

By porting our REST API to something like Swagger, it will also force us to clean the API and make it a bit easier to work with 😃

**Comment by Dmole** [comment]
Also all UI communication to the core should be forced to use the API (or have the HTTP API auto updated from whatever the UI API is using)

For example the new move feature
https://github.com/Tribler/tribler/issues/2901#issuecomment-493388260
I don't see that in the API
https://github.com/Tribler/tribler/blob/master/Tribler/Core/Modules/restapi/downloads_endpoint.py
but maybe I'm looking on the wrong branch or something.

**Comment by qstokkink** [comment]
@Dmole From reverse engineering this unit test:

https://github.com/Tribler/tribler/blob/562b8d86b4dcf4c821b1df5039276734116a6dd3/Tribler/Test/Core/Modules/RestApi/test_downloads_endpoint.py#L487-L499

Moving downloads should be done with an `HTTP PATCH` request to `'downloads/<HEX ENCODED INFOHASH>'` and a request body of:

```
{
"state": "move_storage", 
"dest_dir": <NEW DIRECTORY> 
}
```

**Comment by Dmole** [comment]
Ah it's on devel not master, thanks.

**Comment by heldersepu** [comment]
> What is our documentation's target audience? Developers? Personally, I don't use it. Extensive in-code comments and proper code structure are much more useful to me. Users? Haven't met one who reads the docs. Power users? Maybe. Fresh developers and external contributors? It is hard to say.
> 

@qstokkink  Consider me part of that target audience... What am I? Developer? SRE? Power user? 
But yes the docs is the first place I go when interacting with a REST api, that is all you get on non OpenSource projects, I had a "hard time" putting together a script to queue downloads:
https://github.com/heldersepu/hs-scripts/blob/master/Bash/tribler_queue.sh

**Comment by ichorid** [comment]
As @heldersepu pointed out, the hottest point of our documentation is the REST API because it is, well, *the* interface for Tribler. Remind that we had some discussions about moving to a proper Web API toolkit, like Flask's `asyncio`-based counterpart [Quart](https://hackernoon.com/3x-faster-than-flask-8e89bfbe8e4f). I wonder if it is possible to use Flask's robust Swagger-integration features ([RESTPlus](http://michal.karzynski.pl/blog/2016/06/19/building-beautiful-restful-apis-using-flask-swagger-ui-flask-restplus/)) with Quart...

**Pull request merged**